refactor TyperDotPath (closes #9189, closes #9190)

nadako opened the pull request:
let's see what CI has to say about this, also please review, my brain is fried now so i might've missed something

TODO:
 - [x] check/fix positions

nadako committed:
refactor TyperDotPath (closes #9189, closes #9190)

Simn commented on src/typing/typeload.ml:
```diff
@@ -121,36 +121,72 @@ let find_type_in_current_module_context ctx pack name =
+					with Error (Module_not_found mpath,_) when mpath = path ->
+						raise Not_found
+				in
+				let r = f m ~resume:true in
```
I'm a bit worried that this relies on the `f` callback raising `Not_found`. But that can probably be solved via documentation.

nadako committed:
use correct position for sub-type access

nadako committed:
shut up Issue4720 test for now

this is a separate issue with deprecation warning duplicates that was surfaced by fixing positions in resolved dotpaths

nadako committed:
adapt test position

nadako committed:
improve "sensible error" algorithm

and add a TODO because it should not be needed if we raise proper errors instead of Not_found

nadako committed:
improve error position for the "sensible error"

nadako committed:
fix deprecation warning test
